KEP-3939: Consider terminating pods in job controller (#3940)
kubernetes/enhancements: keps/sig-apps/3939-allow-recreation-when-fully-terminated/README.md

#### Story 1

As a machine learning user, ML frameworks allow scheduling of multiple pods.  
The Job controller does not typically wait for terminating pods to be marked as failed.  
Tensorflow and other ML frameworks may have a requirement that they only want Pods to be started once the other pods are fully terminated.

This case was added due to a bug discovered with running IndexedJobs with Tensorflow.  
See [Jobs create replacement Pods as soon as a Pod is marked for deletion](https://github.com/kubernetes/kubernetes/issues/115844) for more details.

#### Story 2

As a cloud user, users would want to guarantee that the number of pods that are running is exactly the amount that they specify.  
Terminating pods do not relinguish resources so scarce compute resource are still scheduled to those pods.
See [Kueue: Account for terminating pods when doing preemption](https://github.com/kubernetes-sigs/kueue/issues/510) for an example of this.

### Notes/Constraints/Caveats (Optional)

### Risks and Mitigations

One area of contention is how this KEP will work with 3329-retriable-and-non-retriable-failures.  It is important to mention the subtleties here.  

> In 3329, there was a decision to force kubelet to transition pods to failed before marking them as failed.  This is feature toggled guarded by `PodDisruptionCondition`.  

Reviewer: ```suggestion
In 3329, there was a decision to make kubelet transition pods to failed before deleting them.  This is feature toggled guarded by `PodDisruptionCondition`.  
```

Author: done.

> This means that when this feature is turned on we will only mark pods as failed once they are fully terminated.  This means that the pod is still considered running, as opposed to failed.  

Reviewer: ```suggestion
This means that when this feature is turned on, the job controller can count pods as failed once they are fully terminated.  This means that the pod is still considered running, as opposed to failed.  
```

Author: done.

> If `PodDisruptionCondition` is turned off then a pod is marked as failed as soon as it is terminated so it will be counted as failed rather than active.  

Reviewer: ```suggestion
If `PodDisruptionCondition` is turned off then the job controller considers the pod as failed as soon as it is terminating (has a deletion timestamp), because there is no guarantee that the pod will transition to phase=Failed.
```

Author: done.

> To get around this problem we decided to add a field called terminating so we can count separately from active or failed.  

Reviewer 2: I'm not sure I understand how this sentence is related to the sentences above. 

First, indeed when PodDisruptionConditions=true pods are always transitioned to a terminal phase, but this is actually an enabler for this KEP, not an obstacle (because the new semantics of this KEP relies on this assumption). We just need to ensure that the same semantic is applied when the feature gate in this feature is enabled (as described in the paragraph below).

Second (and I guess this is the point that was really intended here), when `PodDisruptionConditions=true`, `JobPodFailurePolicy=true`, and `podFailurePolicy!=nil` (https://github.com/kubernetes/kubernetes/blob/bb67eb5bd237334d6d0343aa382b351002653404/pkg/controller/job/job_controller.go#L1700), then the `terminating` pods are not counted into `.status.failed`, while before they were counted. 

So, the `.status.terminating` field can be useful to clients who need to know the total number of failed and terminating pods at a given moment - previously that number was in `.status.failed`, now it can be computed `.status.failed` + `status.terminating`. So this could justify adding the new field. The new field seems also useful on its own.

Author: Maybe I am being confused by items like https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller.go#L998

Author: Updated.  PTAL

 // ie this means that as soon as pods get marked for deletion they will be recreated
 TerminatingOrFailed RecreatePodsWhen = "TerminatingOrFailed"
 // Only recreate pods when they are marked as failed.
 Failed              RecreatePodsWhen = "Failed"
)
```

```golang
type JobSpec struct{
  ...
 // RecreatePodsWhen specifies when pods should be recreated.
 // TerminatingOrFailed means to recreate when a pod is either terminating or failed
 // TerminatingOrFailed is the default.
 // Failed means to wait until pods are fully terminated or failed before recreating
 // +optional
 RecreatePodsWhen *RecreatePodsWhen
}
```

So we can count terminating pods separately from active or failed we need to include a new field in the JobStatus.

```golang
type JobStatus struct {
  ...
  // Number of terminating pods
  // +optional
  terminating *int32
}
```

We will allow only opt-in behavior for this feature so we will fall back to `TerminatingOrFailed` for `RecreatePodsWhen`.  

### Implementation

The Job Controller filters all pods and classifies the pods as active.  We will append terminating pods to list as part of this KEP.

> There are two cases in the job controller where deletions can happen programatically:

Reviewer 2: TBH I don't understand why the distinction is made here (on pods being deleted by Job controller itself and those deleted by external controllers, such as scheduler during preemption or Kubelet on shutting down a node). 

Can we just say that we provide the new `.status.terminating` field that counts the number of terminating pods (with deletionTimestamp!= nil), but not terminal yet?

Author: I seem to remember someone asking how will this behave with `PodFailurePolicy` or `ActiveDeadlineSeconds` so I made a point to mention it.

Reviewer 2: I see, makes sense, but then mention that "in particular, ...", to make it clear that the distinction isn't important from the design perspective. It is just to provide examples of scenarios covered. 

Also, IMO this does not belong to Design Details -> Implementation. I suggest this could go to the "Notes/Constraints/Caveats" section.

Author: Moved.

## Motivation

Existing Issues:

- [Job Creates Replacement Pods as soon as Pod is marked for deletion](https://github.com/kubernetes/kubernetes/issues/115844)
- [Kueue: Account for terminating pods when doing preemption](https://github.com/kubernetes-sigs/kueue/issues/510)

Many common machine learning frameworks, such as Tensorflow and JAX, require unique pods.  Currently if a pod is killed, a replacement pod is created and can cause undesirable behavior.  This is a rare case but it can provide problems if a job needs to guarantee that the existing pods terminate
before starting new pods.  

In scarce compute environments, these resources can be difficult to obtain so pods can take a long time to find resources and they may only be able to find nodes once the existing pods have been terminated.

If a job is stuck in terminating, it could be possible for autoscaling to kick in and give a new node.  
This is not ideal if you could just wait for the pod to terminating and reuse that node.  

### Goals

- Job controller should allow for flexibility in waiting for pods to be fully terminated before
  creating new ones

> - Job controller will have a new status field where we include the number of terminating pods.

Reviewer: see my comment below on story 2.

> - Job controller will have a new status field where we include the number of terminating pods.

Author: i tried to make this more clear.

### Non-Goals

- Other workload APIs are not included in this proposal.

## Proposal

The Job controller gets a list of active pods.  Active pods usually mean pods that have not been registered for deletion.  
In this KEP, we will consider terminating pods to be separate from active and failed.  
In the job controller, we should include a field that states the number of terminating pods.  

We propose two new API fields:

> 1) A field in Spec that allows for opt-in behavior of whether to wait for terminating pods to finish before recreating.

Reviewer: It could be used for regular Jobs as well, as a means to avoid over-committing clusters (have more running pods than initially planned).
This is somewhat mentioned in story 2, but could be made more explicit.

> 1) A field in Spec that allows for opt-in behavior of whether to wait for terminating pods to finish before recreating.

Author: Tracking of terminating pods is useful for both.  I tried to add story 3 to mention why tracking would be important to both cases.  I don't know if uniqueness would be that important for a NonIndexedJob.

2) A new field in Status for tracking the number of terminating pods.

### User Stories (Optional)

#### Story 1

As a machine learning user, ML frameworks allow scheduling of multiple pods.  
The Job controller does not typically wait for terminating pods to be marked as failed.  
Tensorflow and other ML frameworks may have a requirement that they only want Pods to be started once the other pods are fully terminated.

This case was added due to a bug discovered with running IndexedJobs with Tensorflow.  
See [Jobs create replacement Pods as soon as a Pod is marked for deletion](https://github.com/kubernetes/kubernetes/issues/115844) for more details.

#### Story 2

As a cloud user, users would want to guarantee that the number of pods that are running is exactly the amount that they specify.  
Terminating pods do not relinguish resources so scarce compute resource are still scheduled to those pods.

> See [Kueue: Account for terminating pods when doing preemption](https://github.com/kubernetes-sigs/kueue/issues/510) for an example of this.

Reviewer: This issue is related, but it's not exactly this story.

The use case in Kueue is that we want to be able to track how many pods are still terminating, regardless of whether the user wants to control the amount of running pods.

Consider splitting into a third story

Author: I split into a 3rd story.

### Notes/Constraints/Caveats (Optional)

> A focus of this KEP is for nonprogramatic deletion (ie kubelet eviction, preemption, etc) so we want to mention some other cases where the job controller can start deletion.

Reviewer: not sure what you mean by "non-programmatic". Maybe a better term is just disruptions?

Author: updated.

1) A job is over the `activeDeadlineSeconds` so the child pods are deleted.  

> 2) With `PodFailurePolicy` active and `FailJob` is set as the action, the children pods would be deleted.

Reviewer: ```suggestion
2) With `PodFailurePolicy` active and `FailJob` is set as the action, the pods that are still running would be deleted.
```

### Risks and Mitigations

One area of contention is how this KEP will work with 3329-retriable-and-non-retriable-failures.  It is important to mention the subtleties here.  
In 3329, there was a decision to make kubelet transition pods to failed before deleting them.  This is feature toggled guarded by `PodDisruptionCondition`.  
This means that when this feature is turned on, the job controller can count pods as failed once they are fully terminated.  This means that the pod is still considered running, as opposed to failed.  So this pod would be considered active until it is fully terminated.  
If `PodDisruptionCondition` is turned off then the job controller considers the pod as failed as soon as it is terminating (has a deletion timestamp), because there is no guarantee that the pod will transition to phase=Failed.

> This causes some problems in tracking for active versus failed because we have different behavior based on `PodDisruptionCondition`.  To migitate this, we decided to add a new field called `terminating`.  This will track terminating pods separately from active and failed.  

Reviewer: I agree, the `terminating` field is unrelated here.

The mitigation is rather the paragraph below.

The `terminating` field is just another feature of the KEP.

> This causes some problems in tracking for active versus failed because we have different behavior based on `PodDisruptionCondition`.  To migitate this, we decided to add a new field called `terminating`.  This will track terminating pods separately from active and failed.  

Author: I added a explicit story on why the terminating field could be useful for frameworks like Kueue or any case where people closely watch pods and resources.

> This causes some problems in tracking for active versus failed because we have different behavior based on `PodDisruptionCondition`.  To migitate this, we decided to add a new field called `terminating`.  This will track terminating pods separately from active and failed.  

Reviewer: Thank you for adding the story.

Now, this paragraph can be simplified. You can remove mentions of the `terminating` in this risk, as it's unrelated.

Another issue is described here: https://github.com/kubernetes/enhancements/pull/3940#discussion_r1180777509.  

> If PodDisruptionConditions is disabled, a pod bound to a no-longer-existing node may be stuck in the Running phase. As a consequence, it will never be replaced, so the whole job will be stuck from making progress.  Due to the above issue, we can set phase=Failed when PodDisruptionConditions is enabled OR JobRecreatePodsWhenFailed is enabled. When JobRecreatePodsWhenFailed enabled, but PodDisruptionConditions disabled we would just set the phase, but without adding the condition.  We will need to modify the PodGC (gc_controller.go) in the case of `JobRecreatePodsFailed` being enabled while `PodDisruptionConditions` is disabled.  

Author: Yea I tried to add some test cases.

## Design Details

### Job API Definition

At the JobSpec level, we are adding a new enum field:

```golang
// This field controls when we recreate pods
// Default will be TerminatingOrFailed ie recreate pods when they are failed
// +enum 

> type RecreatePodsWhen string

Author: That would probably fit better with deployment and other apps controllers, right @atiratree ?

> type RecreatePodsWhen string

Author: @alculquicondor consensus on new name?  Not sure if I should change it or not yet.  Maybe @deads2k could weigh in as final vote?

> type RecreatePodsWhen string

Reviewer: whatever I say will likely change with API review, so @deads2k's input will be better.

If we run out of time before enhancements freeze, we can always tweak during the implementation.

> type RecreatePodsWhen string

Reviewer 3: `PodRecreationPolicy` sounds good and yes it fits in well with the other workloads

> type RecreatePodsWhen string

Reviewer 4: PodRecreationPolicy fits better, but the bigger issue is describing the behavior. I'm not really clear on how this is honored alongside podFailurePolicy.  It seems as though when using podFailurePolicy, this field must be Failed and disallow TerminatingOrFailed.

const (
 // This is a field that recreates pods when they are marked as terminating or failed
 // ie this means that as soon as pods get marked for deletion they will be recreated
 TerminatingOrFailed RecreatePodsWhen = "TerminatingOrFailed"
 // Only recreate pods when they are marked as failed.
 Failed              RecreatePodsWhen = "Failed"
)
```

```golang
type JobSpec struct{
  ...
 // RecreatePodsWhen specifies when pods should be recreated.
 // TerminatingOrFailed means to recreate when a pod is either terminating or failed
 // TerminatingOrFailed is the default.
 // Failed means to wait until pods are fully terminated or failed before recreating
 // +optional

>  RecreatePodsWhen *RecreatePodsWhen

Reviewer 5: A suggestion (would be good to double-check with API approvers) how about we make this field part of `PodFailurePolicy`, this seems somewhat related, wdyt?

Reviewer: I tentatively agree.

@deads2k maybe you can review this API as well? Since you are already reviewing a related KEP in Job.

Author: I have tried to think of a rebuttal but not coming up with one.  We originally included this feature as part of the Retriable KEP but we decided to do a separate KEP.  In that case I took that to mean it should be separate from PodFailurePolicy as we wanted to decouple these two features from one another.  As I continue on this, I'm not really sure.  

And we do consider PodRecrationPolicy on failure or not.  It is tangentially related as PodFailurePolicy is mostly on when to mark pods as failed or retry.  Should we move it to `PodFailurePolicy` with `PodRecreationPolicy`?  

@mimowo WDYT?

Reviewer 3: We could consider this field even when creating new pods after the previous ones have succeeded. And not just the failed ones. This could help with the resource usage issue.  The secondary benefit is that we can use the same approach consistently in other workloads.

}
```

So we can count terminating pods separately from active or failed we need to include a new field in the JobStatus.

```golang
type JobStatus struct {
  ...
  // Number of terminating pods
  // +optional

>   terminating *int32

Reviewer 5: That's still debatable, imo.

Author: I hope my updates clear this up.

Reviewer: It can. But that's extra memory to keep the watch for Pods.

Ideally, Kueue can take decisions based entirely on the Job object.

Reviewer: OTOH, I think a `terminating` field is valuable for users. They might be wondering: why is `active` less than `parallelism`? And the answer is: some pods are terminating, which is not currently visible.

}
```

We will allow only opt-in behavior for this feature so we will fall back to `TerminatingOrFailed` for `RecreatePodsWhen`.  

### Implementation

The Job Controller filters all pods and classifies the pods as active.  This list is used to verify that the expected number of active pods matches the reality.  

As part of this KEP, we want to include pods that are also terminating (`DeletionTimestamp != nil`).  
The field `Status.terminating` will include the number of terminating pods.  Once the pods are fully terminated, this field will be unset again because we will obtain the count of active and terminating pods and update the status.

We will update the Status field in the same place as when we update the number of active jobs so it should not require any extra API calls.  

> In cases where pods are terminating outside of the job controller, we will look at the list of pods that a job has.  And we can classify them as terminating and add that to the status field.

Reviewer 5: Can you make the above description a bit more systematic, by explaining in a few steps the proposed changes. I think you covered them all, but they are rather chaotic :sweat_smile:

### Test Plan

<!--
**Note:** *Not required until targeted at a release.*
The goal is to ensure that we don't accept enhancements with inadequate testing.

All code is expected to have adequate tests (eventually with coverage
expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
when drafting this test plan.

[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
-->

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

#### Prerequisite testing updates

<!--
Based on reviewers feedback describe what additional tests need to be added prior
implementing this enhancement to ensure the enhancements have also solid foundations.
-->

##### Unit tests

- `controller_utils`: `April 3rd 2023` - `56.6`

> - `job`: `April 3rd 2023` - `90.4`

Reviewer 5: What unit tests are you planning to add to cover the new functionality?

Author: I tried to add some examples of the tests.  In some cases I already have an implementation in mind but I need to look more into gc_controller and understand if I should have explicit integration tests around this in addition to unit tests.  

I think that this shouldn't hold the KEP review though as that seems to be more of an implementation detail.

##### Integration tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
- <test>: <link to test coverage>
-->

We will add the following integration test for the Job controller:

JobRecreatePodsWhenFailed Feature Toggle On:

  1) Job starts pods that takes a while to terminate
  2) Delete pods
  3) Verify that pod creation only occurs once terminating pods are removed

> We should test the above with the FeatureToggle off also.

Reviewer 5: See my other comment about the interaction with `PodDisruptionConditions` feature.

##### e2e tests

> <!--

Reviewer 5: Not required for alpha, but it would be good to add ideas around planned tests, if you already have some.

Author: Added an example I was using for my implementation POC.

This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
- <test>: <link to test coverage>
-->

### Graduation Criteria

#### Alpha

- Job controller can consider terminating pods as active
- Job controller counts terminating pods in `JobStatus`.
- Unit Tests
- Integration tests

#### Beta

- Address reviews and bug reports from Alpha users
- E2e tests are in Testgrid and linked in KEP
- The feature flag enabled by default

#### GA

- Address reviews and bug reports from Beta users

> - Write a blog post about the feature

Reviewer 5: That is not a hard requirement, I wouldn't place it here explicitly.

Author: Are you saying to remove the entire graduate criteria section?  Or just the mention of the Blog Post?

Reviewer: the mention of the blogpost

> - Graduate e2e tests as conformance tests

Reviewer 5: That happens only after ga, you can't promote to conformance at ga.

- Lock the `JobRecreatePodsWhenFailed` feature-gate
- Declare deprecation of the `JobRecreatePodsWhenFailed` feature-gate in documentation

#### Deprecation

> - Remove feature toggle and code using that.

Reviewer 5: In GA+1:...

Reviewer: GA+2 https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecation

This could potentially make jobs (where pods are terminated) slower due to waiting for terminating pods
to be fully deleted.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes.
<!--
Describe the consequences on existing workloads (e.g., if this is a runtime
feature, can it break the existing applications?).

Feature gates are typically disabled by setting the flag to `false` and
restarting the component. No other changes should be necessary to disable the
feature.

NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
-->

###### What happens if we reenable the feature if it was previously rolled back?

Terminating pods will no longer be tracked so terminating pods will be considered deleted and new pods will be created.

###### Are there any tests for feature enablement/disablement?

Yes. Unit tests will include the fields off/on and verify behavior.

### Rollout, Upgrade and Rollback Planning

<!--
This section must be completed when targeting beta to a release.
-->

#### How can a rollout or rollback fail? Can it impact already running workloads?

<!--

> Try to be as paranoid as possible - e.g., what if some components will restart

Reviewer 5: The interaction with `PodDisruptionConditions` should be clearly explained here.

Author: Did not get to this yet but I will update next week.

mid-rollout?

Be sure to consider highly-available clusters, where, for example,
feature flags will be enabled on some API servers and not others during the
rollout. Similarly, consider large clusters and how enablement/disablement
will rollout across nodes.
-->

#### What specific metrics should inform a rollback?

<!--

> What signals should users be paying attention to when the feature is young

Reviewer 5: This section is a must.

Author: @alculquicondor or @mimowo any suggestions here?  Not exactly sure what metrics we could use or what exists? 

Is sync_duration one that could be used here?  I would expect the sync_duration to take longer if the feature toggle is on.

Reviewer: yeah, in general sync_duration could give some signal about performance regressions, but I think more important is `job_syncs_total`, as it might show an increased number of failures when the feature is enabled.

Those are the generic metrics to use.

More specific to this feature could be `job_pods_finished_total`, compared to the metric for pod creation (is there one? I don't think this is incrementing any metric)

Next, you have to think whether new metrics make sense, but you have to think in an aggregated manner (all jobs, as opposed to per job). Other than pod creation, I don't think we need a new one.

Author: With what @mimowo has suggested I wonder if a avg metric between terminating and creation may make sense?  

I added a note to mention what metrics could signal an issue.

that might indicate a serious problem?
-->

#### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

<!--
Describe manual testing that was done and the outcomes.
Longer term, we may want to require automated upgrade/rollback tests, but we
are missing a bunch of machinery and tooling and can't do that now.
-->

#### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

<!--
Even if applying deprecation policies, they may still surprise some users.
-->

### Monitoring Requirements

<!--
This section must be completed when targeting beta to a release.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->

#### How can an operator determine if the feature is in use by workloads?

<!--

> Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,

Reviewer 5: I'd consider adding a metric or reusing one of the pre-existing ones, if possible.

Author: Will update next week.

Author: I'm not really sure how an operator can determine if this feature is in use by a metric.  I think we could add one for "average time to create pods once terminated started" but not sure if that is a good metric or not.

Reviewer: We don't need a metric for every feature.
But it might be valuable to have a metric for number of pod creations, perhaps with a label about the operation that triggered it (new, recreateImmediate, recreateAfterTerminal)?

checking if there are objects with field X set) may be a last resort. Avoid
logs or events for this purpose.
-->

#### How can someone using this feature know that it is working for their instance?

If a user terminates pods that are controlled by a deployment/job, then we should wait
until the existing pods are terminated before starting new ones.

When feature is turned on, we will also include a `terminating` field in the Job Status.

> We will add e2e test that determine this.

Reviewer 5: There's a separate section for e2e, that needs to explain what kind of tests you're planning to add.

#### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

<!--
This is your opportunity to define what "normal" quality of service looks like
for a feature.

It's impossible to provide comprehensive guidance, but at the very
high level (needs more precise definitions) those may be things like:
  - per-day percentage of API calls finishing with 5XX errors <= 1%
  - 99% percentile over day of absolute value from (job creation time minus expected
    job creation time) for cron job <= 10%
  - 99.9% of /health requests per day finish with 200 code

These goals will help you determine what you need to measure (SLIs) in the next
question.
-->

#### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

NA

#### Are there any missing metrics that would be useful to have to improve observability of this feature?

### Dependencies

> This feature is closely related to the 3329-retriable-and-nonretriable-failures but not sure if that is considered a dependency.

Reviewer 5: There's a high risk of problems, as you pointed out earlier, wrt to these two features clashing. You should call this out.

Author: Added a call out to Risks section.

#### Does this feature depend on any specific services running in the cluster?

No

### Scalability

> Generally, enabling this will slow down job creation if pods take a long time to terminate.  We would wait

Reviewer: ```suggestion
Generally, enabling this will slow down pod creation if pods take a long time to terminate.  We would wait
```

to create new pods until the existing ones are terminated

#### Will enabling / using this feature result in any new API calls?

No

#### Will enabling / using this feature result in introducing new API types?

We add `RecreatePodsWhen` to `JobSpec`.  This is a enum of two values.

> We add `terminating` to `JobStatus`. This is a pointer to type `int32`.

Reviewer 5: Both are new fields, which you describe two questions below. This is explicitly for new types.

Author: Thanks for that. Updated.

#### Will enabling / using this feature result in any new calls to the cloud provider?

No

#### Will enabling / using this feature result in increasing size or count of the existing API objects?

For Job API, we are adding a enum field named `RecreatePodsWhen` which takes either a `TerminateOrFailed` or `Failed`

- API type(s): enum
- Estimated increase in size: 8B

We are also added a status field for tracking terminating pods.

- API type(s): int32
- Estimated increase in size: 4B

#### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

> <!--

Reviewer 5: This is important section, since like you mentioned before the additional wait time for pod termination will slow down the controller.

Author: updated.

Reviewer: I don't think it makes sense to have SLOs for job completion, as this is outside of the control of the job controller. There can be SLOs for job sync (how fast it reacts to events). Although there aren't any official ones today.

Author: I see what you mean.

N/A
<!--
Focus not just on happy cases, but primarily on more pathological cases
(e.g. probes taking a minute instead of milliseconds, failed pods consuming resources, etc.).
If any of the resources can be exhausted, how this is mitigated with the existing limits
(e.g. pods per node) or new limits added by this KEP?

Are there any tests that were run/should be run to understand performance characteristics better
and validate the declared limits?
-->

### Troubleshooting

<!--
This section must be completed when targeting beta to a release.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.

The Troubleshooting section currently serves the `Playbook` role. We may consider
splitting it into a dedicated `Playbook` document (potentially with some monitoring
details). For now, we leave it here.
-->

#### How does this feature react if the API server and/or etcd is unavailable?

No change from existing behavior of the Job controller.

#### What are other known failure modes?

#### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

> - Initial KEP

Reviewer 5: date in format YYYY-MM-DD: description :wink:

## Drawbacks

Enabling this feature may have rollouts become slower.

## Alternatives

> We discussed having this under the PodFailurePolicy but this is a more general idea than the PodFailurePolicy.

Reviewer 5: Can you elaborate more, since I asked this above.

Author: I wrote a response in another place.  I'd like to hear api-review thoughts on this now or @mimowo?

TLDR was that we originally wanted to include it in retriable-kep but it was thought to be better brought up as a separate feature.  I thought that would mean we should decouple it from this API but I realize that I am not so sure anymore.

Reviewer 2: I think the reason to split was to avoid the snowball effect when more loosely related goals are added to the pod failure policy KEP. In particular it introduces a new API field, so requires at least two releases (alpha and beta), which could prolong graduation of the pod failure policy unnecessarily.

Tensorflow and other ML frameworks may have a requirement that they only want Pods to be started once the other pods are fully terminated.

This case was added due to a bug discovered with running IndexedJobs with Tensorflow.  
See [Jobs create replacement Pods as soon as a Pod is marked for deletion](https://github.com/kubernetes/kubernetes/issues/115844) for more details.

#### Story 2

As a cloud user, users would want to guarantee that the number of pods that are running is exactly the amount that they specify.  
Terminating pods do not relinguish resources so scarce compute resource are still scheduled to those pods.
See [Kueue: Account for terminating pods when doing preemption](https://github.com/kubernetes-sigs/kueue/issues/510) for an example of this.

### Notes/Constraints/Caveats (Optional)

### Risks and Mitigations

One area of contention is how this KEP will work with 3329-retriable-and-non-retriable-failures.  It is important to mention the subtleties here.  

In 3329, there was a decision to force kubelet to transition pods to failed before marking them as failed.  This is feature toggled guarded by `PodDisruptionCondition`.  

> This means that when this feature is turned on we will only mark pods as failed once they are fully terminated.  This means that the pod is still considered active.  

Reviewer: ```suggestion
This means that when this feature is turned on we will only mark pods as failed once they are fully terminated.  This means that the pod is still considered running, as opposed to failed.  
```

> If this feature is turned off then a pod is marked as failed as soon as it is terminated so it will be counted as failed rather than active.  

Reviewer: ```suggestion
If `PodDisruptionCondition` is turned off then a pod is marked as failed as soon as it is terminated so it will be counted as failed rather than active.  
```

> To get around this problem we decided to add a field called terminating so we can count separately from active or failed.  

Reviewer: This sentence seems out of context?

Author: Updated.  This is supposed to be an effect of the Retriable pods KEP.

Reviewer: I'm not sure I understand. The feature doesn't currently exist. It's not clear if this sentence is referring to what this KEP is going to change or pre-existing behavior.

> Another issue is described here: https://github.com/kubernetes/enhancements/pull/3940#discussion_r1180777509.  

Author: Done.

If PodDisruptionConditions is disabled, a pod bound to a no-longer-existing node may be stuck in the Running phase. As a consequence, it will never be replaced, so the whole job will be stuck from making progress.

> Due to the above issue, we can set phase=Failed when PodDisruptionConditions is enabled OR JobRecreatePodsWhenFailed is enabled. When JobRecreatePodsWhenFailed enabled, but PodDisruptionConditions disabled we would just set the phase, but without adding the condition.

Reviewer: ```suggestion
To prevent the above issue, we can set phase=Failed when PodDisruptionConditions is enabled OR JobRecreatePodsWhenFailed is enabled. When JobRecreatePodsWhenFailed enabled, but PodDisruptionConditions disabled we would just set the phase, but without adding the condition.
```

also, where will we do this? just pod GC, kubelet, scheduler, etc?

Author: @mimowo would know best but I think maybe kubelet?

Reviewer 2: We only need to this is PodGC (gc_controller.go). The changes in kubelet to make sure pods transition to failed phase are done unconditionally (not behind a feature gate).

Author: updated with a mention to update PodGC in this case.

Tensorflow and other ML frameworks may have a requirement that they only want Pods to be started once the other pods are fully terminated.

This case was added due to a bug discovered with running IndexedJobs with Tensorflow.  
See [Jobs create replacement Pods as soon as a Pod is marked for deletion](https://github.com/kubernetes/kubernetes/issues/115844) for more details.

#### Story 2

As a cloud user, users would want to guarantee that the number of pods that are running is exactly the amount that they specify.  
Terminating pods do not relinguish resources so scarce compute resource are still scheduled to those pods.

#### Story 3

The use case in Kueue is that we want to be able to track how many pods are still terminating, regardless of whether the user wants to control the amount of running pods.  Kueue wants to track the number of pods using resources and terminating pods still utilize resources.  Without having a way to track terminating pods, it is not possible to know if a terminating pod is still using resources.

See [Kueue: Account for terminating pods when doing preemption](https://github.com/kubernetes-sigs/kueue/issues/510) for an example of this.

> 

Reviewer 2: Not sure if this deserves a story, but I think debugging a failed job can be less confusing with this feature. In particular, it is possible that some pods that eventually terminate as Success, are incrementing the `Failed` counter. I guess this might be confusing for some users trying to understand what happened. WDYT? @alculquicondor @soltysh

Reviewer: How does this feature help in that case?

Reviewer 2: The feature actually fixes the problem, because it awaits with counting as Failed or Succeeded, until it is really terminal. 

It may not be a story, but worth mentioning somewhere. OTOH it could be a story for debugging, as the number of successful or failed pods not agreeing with the Job counter might be misleading.

## Design Details

### Job API Definition

At the JobSpec level, we are adding a new enum field:

```golang
// This field controls when we recreate pods
// Default will be TerminatingOrFailed ie recreate pods when they are failed
// +enum 
type RecreatePodsWhen string
const (
 // This is a field that recreates pods when they are marked as terminating or failed
 // ie this means that as soon as pods get marked for deletion they will be recreated
 TerminatingOrFailed RecreatePodsWhen = "TerminatingOrFailed"
 // Only recreate pods when they are marked as failed.

>  Failed              RecreatePodsWhen = "Failed"

Reviewer 2: I think we should also discuss how this relates to the backoff delay. Currently, we replace a failed pod with a delay of 10s, IIUC for deleted pods this is essentially from the moment of deletion. On the other hand the typical graceful period is 30s, so the pod is recreated before the old pod deletion. However, once we wait until the old pod is terminal, waiting 10s may introduce a significant delay. Maybe we should instead use 0s, 10s, 20s, 40s... WDYT @alculquicondor @soltysh ?

Reviewer: Good point. Added to notes.

)
```

```golang
type JobSpec struct{
  ...
 // RecreatePodsWhen specifies when pods should be recreated.
 // TerminatingOrFailed means to recreate when a pod is either terminating or failed

>  // TerminatingOrFailed is the default.

Reviewer 2: I'm wondering if we should consider changing the default, leaving `TerminatingOrFailed` to allow for the old semantics. I think the old mode is also confusing for users who may see the `Failed` counter incremented even if the pod eventually terminates with Success. See also the comment about a debugging story. WDYT @alculquicondor @soltysh ?

Reviewer 4: I agree that it's confusing, *but* it's the existing behavior. Make a note of this for "things you'd like to change in a v2", but keep the compatible behavior in v1.

Reviewer: Adding a paragraph about v2 in the Notes section.

Another issue is described [here](https://github.com/kubernetes/enhancements/pull/3940#discussion_r1180777509).
If PodDisruptionConditions is disabled, a pod bound to a no-longer-existing node may be stuck in the Running phase. As a consequence, it will never be replaced, so the whole job will be stuck from making progress.  Due to the above issue, we can set phase=Failed when PodDisruptionConditions is enabled OR JobRecreatePodsWhenFailed is enabled. When JobRecreatePodsWhenFailed enabled, but PodDisruptionConditions disabled we would just set the phase, but without adding the condition.  We will need to modify the PodGC (gc_controller.go) in the case of `JobRecreatePodsFailed` being enabled while `PodDisruptionConditions` is disabled.  

## Design Details

### Job API Definition

At the JobSpec level, we are adding a new enum field:

```golang
// This field controls when we recreate pods
// Default will be TerminatingOrFailed ie recreate pods when they are failed
// +enum 
type RecreatePodsWhen string
const (
 // This is a field that recreates pods when they are marked as terminating or failed
 // ie this means that as soon as pods get marked for deletion they will be recreated

>  TerminatingOrFailed RecreatePodsWhen = "TerminatingOrFailed"

Reviewer 4: How does this interact with the `PodFailurePolicy` today?  is it
1. pod gets a deletiontimestamp and then say a grace period of 10 minutes to shutdown.
2. job controller marks as failed and creates new pod
3. PodFailurePolicy somehow rewrites the failure after 10 minutes?

Reviewer: When using PodFailurePolicy, we still need to wait for the pod to terminate, because the policy might require an exit code. In other words, the current behavior is equivalent `recreatePodsWhen: Failed` when the user sets a `podFailurePolicy`.

Reviewer 4: > Edit: The behavior is that the job controller immediately creates a replacement Pod, even though the previous pod is not counted as "failed" yet.

I'm still confused then. Does this mean that the `PodFailurePolicy` is ineffective in the default case if the termination grace period is long enough for the controller to take action before the pod has an exit code?

Reviewer: To some degree.
The job controller would still wait for the Pod to finish, and take action based on it.
If the action is "FailJob", then the job controller would still fail the job and delete the replacement pod that it created earlier.

Reviewer 4: I create this

```yaml
spec:
  backoffLimit: 0
  recreatePodsWhen: nil
  podFailurePolicy:
    rules:
    - action: Ignore
      onExitCodes: 
        operator: In
        values:
        - 127
# podspec with 10 minute graceful termination
```

With defaulting, I get back
```yaml
spec:
  backoffLimit: 0
  recreatePodsWhen: TerminatingOrFailed
  podFailurePolicy:
    rules:
    - action: Ignore
      onExitCodes: 
        operator: In
        values:
        - 127
# podspec with 10 minute graceful termination
```

If the pod is deleted (deletionTimestamp != nil), then the pod is clearly terminating, but it may still exit with code 127, we don't know yet. If the pod is considered failed by the job, then no replacement pod should be created and the job should be failed.  But if the pod exits with 127, then the failure should be ignored and a new pod created.  If the controller honors the TerminatingOrFailed, then the doc says it will be created, but this would violate the backoffLimit.  If the controller honors the podFailurePolicy and waits for the exit code, then the recreatePodsWhen was not honored.

What is proposed to happen?

Reviewer: This is one of those tricky scenarios. Because you are using `podFailurePolicy`, the Pod is not counted towards `active` nor `failed`. Under this perspective, the pod recreation is legit.

If you use the new `recreatePodsWhen: Failed`, then we will:
- respect backoffLimit based on `failed`.
- track Pods that are "active or terminating". We would hold recreation based on this tracker, independently of the value of `active`.

One alternative is to consider this a bug of `podFailurePolicy` (still in Beta) and always track "active or terminating". Then this means that when using `podFailurePolicy` the default is `recreatePodsWhen: Failed` and you cannot use `recreatePodsWhen: TerminatingOrFailed`.

 // Only recreate pods when they are marked as failed.
 Failed              RecreatePodsWhen = "Failed"
)
```

```golang
type JobSpec struct{
  ...
 // RecreatePodsWhen specifies when pods should be recreated.
 // TerminatingOrFailed means to recreate when a pod is either terminating or failed
 // TerminatingOrFailed is the default.
 // Failed means to wait until pods are fully terminated or failed before recreating
 // +optional
 RecreatePodsWhen *RecreatePodsWhen
}
```

So we can count terminating pods separately from active or failed we need to include a new field in the JobStatus.

```golang
type JobStatus struct {
  ...
  // Number of terminating pods
  // +optional
  terminating *int32
}
```

We will allow only opt-in behavior for this feature so we will fall back to `TerminatingOrFailed` for `RecreatePodsWhen`.  

### Implementation

As part of this KEP, we want to include pods that are also terminating (`DeletionTimestamp != nil`).  

The following algorithm could be used:

1) Count the number of pods that are active

> 2) If `Failed` and feature is on, then we will count the number of terminating pods

Reviewer 4: Why are terminating pods counted under `Failed` instead of under `TerminatingOrFailed`?  Typo?

Reviewer: Not a typo, but we should always calculate terminating pods. Adjusted the algorithm.
